packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1984,6 +2010,22 @@ export class KeyringController extends BaseController<
       keyrings: this.state.keyrings,
     };
   }
+
+  async #withVaultLock<T>(
+    fn: ({
+      releaseLock,
+    }: {
+      releaseLock: MutexInterface.Releaser;
+    }) => Promise<T>,
+  ): Promise<T> {
+    const releaseLock = await this.#vaultOperationMutex.acquire();
+
+    try {
+      return fn({ releaseLock });
+    } finally {
+      releaseLock();
+    }
+  }
 }
```

Comment:
unfortunately this cannot be used with `createNewVaultAndKeychain` and `createNewVaultAndRestore` because it would cause a deadlock, since both initialization operations use `persistAllKeyrings` (which would be behind the same mutex)